atc/db/migration/migrations/1570709099_add_varsources_to_pipelines.up.sql

```diff
@@ -0,0 +1,4 @@
+BEGIN;
+  ALTER TABLE pipelines ADD COLUMN var_sources text,
```

Comment:
Can we also consider my original comment about moving these into their own table. I know it's not a big difference now, but if we ever want to think about de-duplicate credential managers across pipelines (similar to global resources), it will be much easier if these are already split out.

Reply:
I think they're already de-duped across pipelines via the var sources pool; I'm not really sure if putting them in a separate table buys us anything at that point.

Reply:
oh i see now. Each ATC doing their own in memory deduplication and running their own auth loop seems kinda scary to me.

Reply:
I'm not too familiar with how the credential managers work, but I'm guessing the vault cred manager already runs 1 auth loop per atc?

Reply:
Yep - one per ATC today.

In this PR they're "GC'd" from memory after 5 minutes of inactivity, at least.

I think it kind of makes sense since you'd want some form of caching of active clients anyway, so you're not logging in every single time and hammering the credential manager. Most pipelines will be continuously using their credential manager for periodic checking anyway.

The memory usage + active auth loops concern is valid, though it seems kind of tricky to do anything else. If we had a `var_sources` table we could free up the memory use and cache the currently-active credential there instead so we don't have a bunch of auth loops. The downside is then we'd be persisting credentials into the database... which we can encrypt, but I wonder how people would feel about that.

It seems like this `VarSourcePool` object would be a pretty easy place to attempt this sort of thing in the future, which is cool at least!

Reply:
I've actually been thinking about what it might look like to run all third party integrations (auth, creds, metrics) as containers on workers. Where each one could be represented by your new `prototype` concept and come with some pre baked actions.

Reply:
As a user, I would keep secrets in memory only.